lib/proxy/peer/quic/server.go

```diff
@@ -70,31 +84,395 @@ func (c *ServerConfig) checkAndSetDefaults() error {
+		return
+	}
+	if reqLen >= quicMaxMessageSize {
+		log.WarnContext(conn.Context(), "received oversized request", "request_len", reqLen)
+		return
+	}
+	reqBuf := make([]byte, reqLen)
+	if _, err := io.ReadFull(stream, reqBuf); err != nil {
+		log.DebugContext(conn.Context(), "failed to read request", "error", err)
+		return
+	}
+	stream.SetReadDeadline(time.Time{})
+
+	req := new(quicpeeringv1a.DialRequest)
+	if err := proto.Unmarshal(reqBuf, req); err != nil {
+		log.WarnContext(conn.Context(), "failed to unmarshal request", "error", err)
+		return
+	}
+
+	if req.GetPing() {
+		log.DebugContext(conn.Context(),
+			"received ping request",
+			"nonce", req.GetNonce(),
+			"request_timestamp", req.GetTimestamp().AsTime(),
+		)
+		select {
+		case <-conn.HandshakeComplete():
+		case <-conn.Context().Done():
+			log.DebugContext(conn.Context(),
+				"handshake failure or connection loss after receiving ping request",
+				"error", context.Cause(conn.Context()),
+			)
+			return
+		}
+		stream.CancelRead(0)
+		if _, err := stream.Write([]byte(dialResponseOK)); err != nil {
+			return
+		}
+		_ = stream.Close()
+		closedStream = true
+		return
+	}
+
+	if requestTimestamp := req.GetTimestamp().AsTime(); time.Since(requestTimestamp).Abs() > quicTimestampGraceWindow {
+		log.WarnContext(conn.Context(),
+			"dial request has out of sync timestamp, 0-RTT performance will be impacted",
+			"request_timestamp", requestTimestamp,
+		)
+		select {
+		case <-conn.HandshakeComplete():
+		case <-conn.Context().Done():
+			// logging this at warn level because it should be very atypical to
+			// begin with, and it might be a symptom of a malicious actor
+			// interfering with the connection
+			log.WarnContext(conn.Context(),
+				"handshake failure or connection loss after receiving dial request with out of sync timestamp",
+				"error", context.Cause(conn.Context()),
+			)
+			return
+		}
+	}
+
+	// a replayed request is always wrong even after a full handshake, the
+	// replay might've happened before the legitimate request
+	if !s.replayStore.add(req.GetNonce(), time.Now()) {
+		log.ErrorContext(conn.Context(), "request is reusing a nonce, rejecting", "nonce", req.GetNonce())
+		sendErr(trace.BadParameter("reused or invalid nonce"))
+		return
+	}
+
+	_, clusterName, ok := strings.Cut(req.GetTargetHostId(), ".")
+	if !ok {
+		sendErr(trace.BadParameter("server_id %q is missing cluster information", req.GetTargetHostId()))
+		return
+	}
+
+	nodeConn, err := s.dialer.Dial(clusterName, peerdial.DialParams{
+		From: &utils.NetAddr{
+			Addr:        req.GetSource().GetAddr(),
+			AddrNetwork: req.GetSource().GetNetwork(),
+		},
+		To: &utils.NetAddr{
+			Addr:        req.GetDestination().GetAddr(),
+			AddrNetwork: req.GetDestination().GetNetwork(),
+		},
+		ServerID: req.GetTargetHostId(),
+		ConnType: types.TunnelType(req.GetConnectionType()),
+	})
+	if err != nil {
+		sendErr(err)
+		return
+	}
+
+	var eg errgroup.Group
```

Comment:
Set the group limit to 2, to be explicit/safe.

Should we use the group context to cancel in-flight goroutines, or is that redundant? This probably deserves a brief comment.

Reply:
We call Go twice in succession and then immediately call Wait, we don't move or reuse the errgroup, and there's no errgroup context, I'm not entirely sure what the comment should be.